bigtable/bigtable-proxy/src/main/java/com/google/cloud/bigtable/examples/proxy/channelpool/DataChannel.java

```diff
@@ -148,6 +151,58 @@ private void warm() {
     }
   }
 
+  private ListenableFuture<PingAndWarmResponse> sendPingAndWarm(PingAndWarmRequest request) {
+    CallLabels callLabels =
+        CallLabels.create(
+            BigtableGrpc.getPingAndWarmMethod(),
+            Optional.of("bigtableproxy"),
+            Optional.of(request.getName()),
+            Optional.of(request.getAppProfileId()));
+    Tracer tracer = new Tracer(metrics, callLabels);
+
+    CallOptions callOptions =
+        CallOptions.DEFAULT
+            .withCallCredentials(callCredentials)
+            .withDeadline(Deadline.after(1, TimeUnit.MINUTES));
+    callOptions = tracer.injectIntoCallOptions(callOptions);
+
+    ClientCall<PingAndWarmRequest, PingAndWarmResponse> call =
+        inner.newCall(BigtableGrpc.getPingAndWarmMethod(), callOptions);
+
+    Metadata metadata = new Metadata();
+    metadata.put(
+        CallLabels.REQUEST_PARAMS,
+        String.format(
+            "name=projects/%s/instances/%s",
+            URLEncoder.encode(request.getName(), StandardCharsets.UTF_8),
+            URLEncoder.encode(request.getAppProfileId(), StandardCharsets.UTF_8)));
+
+    SettableFuture<PingAndWarmResponse> f = SettableFuture.create();
+    call.start(
+        new Listener<>() {
+          @Override
+          public void onMessage(PingAndWarmResponse response) {
+            f.set(response);
+          }
+
+          @Override
+          public void onClose(Status status, Metadata trailers) {
+            tracer.onCallFinished(status);
+
+            if (status.isOk()) {
+              f.setException(new IllegalStateException("PingAndWarm was missing a response"));
+            } else {
+              f.setException(status.asRuntimeException());
+            }
+          }
+        },
+        metadata);
+    call.sendMessage(request);
+    call.request(Integer.MAX_VALUE);
```

Comment:
why is this not request(1)?

Reply:
Good catch, its meant to help me detect when PingAndWarm is misbehaving. But I forgot to add the check until you pointed it out